site/documentation/how-does-opa-work/index.md

```diff
@@ -0,0 +1,512 @@
+
+```http
+PUT https://example.com/v1/policies/exempli-gratia HTTP/1.1
+Content-Type: text/plain
+```
+
+```ruby
+package opa.examples
+
+import data.servers
+import data.networks
+import data.ports
+
+violations[server] :-
+    server = servers[_],
+    server.protocols[_] = "http",
+    public_servers[server] = true
+
+public_servers[server] :-
+    server = servers[_],
+    server.ports[_] = ports[i].id,
+    ports[i].networks[_] = networks[j].id,
+    networks[j].public = true
+```
+
+A policy file must contain a single package declaration, which defines the path to the policy module and its rules (for example, data.opa.examples.violations – see The data Document for more information about accessing nested documents). The policy name itself (in this case, “exempli-gratia”) is only used to identify policies for file management purposes; it is not used otherwise.
+
+### Rules and Virtual Documents
+
+In contrast to base documents, virtual documents embody the results of evaluating the rules included in policy modules. Virtual documents are computed when users publish new policy modules, update existing modules, run queries, and when any relevant base document is published or updated. Rules allow policy authors to write questions with yes-no answers (that is, predicates) and to generate structured values from raw data found in base documents as well as from intermediate data found in other virtual documents.
+
+### The data Document
+
+OPA nests all documents within a built-in root document named data. All documents, whether pushed by your service or computed by OPA as needed, are contained within the data document.
+
+{% img '{{assets["data-model-logical.svg"].logical_path}}' width:'720' %}
+
+```json
+{
+  "data": {
+    "servers": [...],
+    "ports": [...],
+    "networks": [...],
+    "opa": {
+      "examples": {
+        "violations": [...],
+        "public_servers": [...]
+      }
+    }
+  }
+}
+```
+
+As a result, any document, base or virtual, can be accessed hierarchically starting from the root data node – either as an identifier:
+
+```ruby
+import data.servers                            # Base document
+import data.opa.examples.violations            # Virtual document
+```
+
+or as a URI component in an HTTP request:
+
+```http
+GET https://example.com/v1/data/servers HTTP/1.1
+```
+
+```http
+GET https://example.com/v1/data/opa/examples/violations HTTP/1.1
+```
+
+### Putting It All Together
+
+Let’s take a look at some documents representing the state of a hypothetical service and a policy module that uses this data. The following documents describe a set of servers, the protocols they use, the ports they open, and the networks those ports are connected to.
+
+```json
```

Comment:
Just for future reference, I think a small illustration of servers/nics/networks could be useful here.